### PR TITLE
8313248: C2: setScopedValueCache intrinsic exposes nullptr pre-values to store barriers

### PR DESCRIPTION
**Pull request by shipilev**
See the bug for investigation breadcrumbs. The root cause for failures seen with Shenandoah seem to be as follows.

The setter (`setScopedValueCache`) intrinsic passes `val_type` of `_gvn.type(arr)`, which is `narrowoop: java/lang/Object *[int:32] (java/lang/Cloneable,java/io/Serializable):NotNull:exact *`, derived from the `argument(0)`, and thus implies non-nullity.

So when Shenandoah's SATB barrier loads the `pre_val`, it folds the null-check, assuming the `pre_val` is not null, due to `val_type`. This passes `nullptr` to SATB queues or slowpath, and we crash in either queue filtering or barrier code that does not expect nullptrs on SATB paths. The getter (`scopedValueCache`) constructs the `objects_type` explicitly to imply the value can be null. I think we should do the same for setter, since it can hide the "getter" from SATB barrier inside of it. 

Arguably, it is a landmine that GC barriers assume the `val_type` is the type of both stored value and the pre-value read from memory. So the non-null-ness derived for stored value gets used to reason for non-null-ness for pre-value. We can explore the solutions to that generic problem after we plug this leak. Other `access_store_at` uses in C2 intrinsics seem to only operate on thread fields that are not null, so the are not susceptible to this problem. `scopedValueCache` is a notable exception of lazily initialized thread OopHandle accessed from C2.

I think G1 SATB barriers have the same problem, but I have not tried to reproduce the failure very hard there. (It would, AFAIU, require writing the test which does G1 concurrent marks, not just young GCs.)

Attn @theRealAph ;)

Additional testing:
 - [x] Linux x86_64 fastdebug, 10+ iterations of `java/lang/ScopedValue/StressStackOverflow.java` with Shenandoah
 - [x] Linux x86_64 fastdebug, `hotspot_loom jdk_loom` with Shenandoah
 - [x] Linux x86_64 fastdebug, `hotspot_loom jdk_loom` with G1
 - [x] Linux AArch64 fastdebug, `tier1 tier2 tier3`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313248](https://bugs.openjdk.org/browse/JDK-8313248): C2: setScopedValueCache intrinsic exposes nullptr pre-values to store barriers (**Bug** - P2)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [ff3ad44b](https://git.openjdk.org/jdk/pull/15105/files/ff3ad44bd632c22c2cffa0fee52a7cf437934126)
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15105/head:pull/15105` \
`$ git checkout pull/15105`

Update a local copy of the PR: \
`$ git checkout pull/15105` \
`$ git pull https://git.openjdk.org/jdk.git pull/15105/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15105`

View PR using the GUI difftool: \
`$ git pr show -t 15105`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15105.diff">https://git.openjdk.org/jdk/pull/15105.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15105#issuecomment-1660211918)

### CODE REVIEW

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -3586,12 +3586,19 @@ bool LibraryCallKit::inline_native_setCurrentThread() {
   return true;
 }
 
-Node* LibraryCallKit::scopedValueCache_helper() {
+const Type* LibraryCallKit::scopedValueCache_type() {
   ciKlass *objects_klass = ciObjArrayKlass::make(env()->Object_klass());
   const TypeOopPtr *etype = TypeOopPtr::make_from_klass(env()->Object_klass());
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:3591]:
```suggestion
  ciKlass* objects_klass = ciObjArrayKlass::make(env()->Object_klass());
  const TypeOopPtr* etype = TypeOopPtr::make_from_klass(env()->Object_klass());
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/opto/library_call.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/opto/library_call.cpp]: Done so, thanks.

**Comment by shipilev** [comment]
Thanks! Any other reviewers for this one?

**Review by rkennke — APPROVED** [review]
Looks good. Thank you!

**Comment by shipilev** [comment]
All right, thanks for reviews! I am integrating to unbreak Shenandoah/G1 with Loom.

/integrate

**Pull request closed without merging**